### DISCUSSION TITLE
CSLA 6: Unable to use csla objects in static ObjectAuthorizationRules method

### DISCUSSION THREAD

**Discussion by JacoJordaan** [MarimerLLC/csla#2790] [Questions]
In my CSLA 4/5 classes I have been adding object type rules from data retrieved from the database.

The static `AddObjectAuthorizationRules`method calls relevant csla objects to retrieve the required permission data from the database. It then constructs and adds the authorisation business rules.

In CSLA 6 you can't create csla objects from a static class. How would this be accomplished in CSLA 6?

If I was able to reference the DataPortal from the static class then this should be possible.
Or if it was in some way possible to reference the service provider then I can call the relevant csla factory classes.

Any ideas?

**Comment by rockfordlhotka** [comment]
Which type of app are you building? There are some workarounds for winforms/wpf/uwp. If you are in aspnetcore/blazor there are less easy workarounds.

**Reply by JacoJordaan** [reply]
I am using Blazor. I would like to try any suggested workarounds.

**Comment by rockfordlhotka** [comment]
In Blazor, the most direct solution is to inject an `IDataPortalFactory` in your page, then provide that to the static factory method so, inside the factory method, you can use the `IDataPortalFactory` to get an instance of the data portal as needed.

```c#
@inject IDataPortalFactory factory;

@code
{
  public async void MyAsyncMethod()
  {
    var obj = await MyType.GetMyType(factory, 123);
  }
}
```
then

```c#
[Serializable]
public class MyType : BusinessBase<MyType>
{
  public static async Task<MyType> GetMyType(IDataPortalFactory factory, int id)
  {
    return await factory.GetPortal<MyType>().FetchAsync(id);
  }
}
```

**Reply by JacoJordaan** [reply]
Ok, I understand this will work with normal business objects. You can pass the DataPortalFactory to the static async method. 
On the other hand, the `AddObjectAuthorisationRules` static method is called by csla itself and it does not allow for any parameters at the moment.
What about allowing an optional IDataPortalFactory parameter for the AddObjectAuthorisationRules static method?
Something like this:
````
[ObjectAuthorizationRules]
private static void AddObjectAuthorizationRules(IDataPortalFactory factory)
{
    
}
````
I am not sure this will work, since the coder has no control to pass in the factory object.
It needs to be passed in by the csla framework and the coder needs to be able to use it.
Your thoughts on this?

**Reply by JacoJordaan** [reply]
I am just thinking about this and I see there are now a GetDataPortal method on the BusinessBase class. What about a protected static GetDataPortal method that will be callable from within the `private static void AddObjectAuthorizationRules()` method?

**Reply by rockfordlhotka** [reply]
It is not possible to access services like the data portal or ApplicationContext from a static method. That reality has been the driver for most of the big changes in CSLA 6 (combined with the fact that ApplicationContext must be a service to work with Blazor).

I'll have to see if it is possible to provide services to the AddObjectAuthorizationRules method. Though that method is static, it is actually being invoked by CSLA, and hopefully not from other static methods!

**Comment by JacoJordaan** [comment]
Thanks Rocky, I thought that it would not be possible to access the services from the static method. I am hopeful that it will be possible to provide services to the AddObjectAuthorizationRules method.

**Reply by rockfordlhotka** [reply]
I opened an issue for this and have started exploring the problem.

It will be a breaking change, because I _think_ what will happen is that we'll pass in some sort of context parameter, and you'll call the `AddRule` method on that context, not as a static like today.

It looks like some fairly deep surgery will be required on the existing codebase to get this working.

I am glad you caught this oversight!!

**Reply by JacoJordaan** [reply]
This sounds promising. I am glad it has been caught while still in the process with breaking changes. Better now than after the final release.

**Reply by mtavares628** [reply]
@rockfordlhotka I just want to bring up the possibility of async authorization rules again (#2643) since this seems like it goes hand in hand with this.  If the auth rule needs the data portal, a synchronous call isn't going to work in blazor.  This would actually be a huge help for my object auth rules which need to hit the database.  Let me know if I'm off base, and if this is different than what @JacoJordaan  is asking.

**Comment by rockfordlhotka** [comment]
Ongoing discussion in #2796 - @JacoJordaan your input will be valuable.

**Comment by dinparacha1** [comment]
following is working solution that is working in my environment        

 public static void AddObjectAuthorizationRules([Inject] AddObjectAuthorizationRulesContext authContext)
        {
            var TableRightsPortal = authContext.DataPortalFactory.GetPortal<TableRightsList>();

             TableRightsList.Factory objTableRights = new TableRightsList.Factory(TableRightsPortal);
             objTableRights.ApplyAuth<MasjidEdit>("MyTableName");

}

**Comment by dinparacha1** [comment]
and here is my TableRightsList Businesss List Class

using System;
using System.Linq;
using Csla;
using System.Collections.Generic;
using System.Threading.Tasks;
using Security.DalEfCore;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using Security.Dal;
using Csla.Server;


namespace Security.Library
{
    [Serializable()]
    public class TableRightsList : ReadOnlyListBase<TableRightsList, TableRightsInfo>
    {

        public static List<string> strRoles = new List<string>();

        public List<string> GetRoles()
        {
            return strRoles;
        }

        public void RemoveChild(string RoleName)
        {
            var iro = IsReadOnly;
            IsReadOnly = false;
            try
            {
                var item = this.Where(r => r.TableName == RoleName).FirstOrDefault();
                if (item != null)
                {
                    var index = IndexOf(item);
                    Remove(item);
                }
            }
            finally
            {
                IsReadOnly = iro;
            }
        }

        public class Factory
        {
            private IDataPortal<TableRightsList> _dataPortal;

            public Factory(IDataPortal<TableRightsList> dataPortal)
            {
                _dataPortal = dataPortal;
            }

            public async Task<TableRightsList> GetTableRightsListAsync()
            {
                return await _dataPortal.FetchAsync();
            }

            public TableRightsList GetTableRightsList()
            {
                return _dataPortal.Fetch();
            }

            // public static TableRightsList GetTableRightForTableRolesList(string Table_Name, string Role_Name)
            // {
            //   return DataPortal.Fetch<TableRightsList>(Table_Name, Role_Name);
            // }
            public TableRightsList GetTableRightsForTableList(string Table_Name)
            {
                return _dataPortal.Fetch(Table_Name, null);
            }

            public TableRightsList GetTableRightForRolesList(string Role_Name)
            {
                return _dataPortal.Fetch(null, Role_Name);
            }



            public async Task<List<string>> FetchViewRoles(string tableName)
            {

                await _dataPortal.FetchAsync(tableName, 1, 0, 0, 0);

                return strRoles;
            }

            public async Task<List<string>> FetchAddRoles(string tableName)
            {
                await _dataPortal.FetchAsync(tableName, 0, 1, 0, 0);

                return strRoles;
            }

            public async Task<List<string>> FetchDeleteRoles(string tableName)
            {
                await _dataPortal.FetchAsync(tableName, 0, 0, 1, 0);

                return strRoles;
            }

            public async Task<List<string>> FetchEditRoles(string tableName)
            {
                await _dataPortal.FetchAsync(tableName, 0, 0, 0, 1);

                return strRoles;
            }

            public void ApplyAuth<T>(string p_Module_Name)//,[Inject] ITableRightsDal dal)
            {
                // DataPortal<TableRightsList> tableRightsListPortal = new DataPortal<TableRightsList>();

                //List<string> strRoles1 = new List<string>();

                // strRoles =await  FetchViewRoles(p_Module_Name);
                // strRoles = await tableRightsListPortal.FetchViewRoles(p_Module_Name);
                // strRoles = objList.GetRoles();
                // strRoles = await DataPortal<TableRightsList>.FetchAsync(p_Module_Name); //(p_Module_Name, 1, 0, 0, 0);


                //Load Add rights
                TableRightsList objList = _dataPortal.Fetch(p_Module_Name, 0, 1, 0, 0);
                Csla.Rules.BusinessRules.AddRule(
                              typeof(T),
                              new Csla.Rules.CommonRules.IsInRole(
                                Csla.Rules.AuthorizationActions.CreateObject,
                               strRoles));



                //Load View rights
                if (p_Module_Name != "Users") { 
                 objList = _dataPortal.Fetch(p_Module_Name, 1, 0, 0, 0);
                Csla.Rules.BusinessRules.AddRule(
                              typeof(T),
                              new Csla.Rules.CommonRules.IsInRole(
                                Csla.Rules.AuthorizationActions.GetObject,
                               strRoles));
				}

                


                //Load Delete rights
                 objList = _dataPortal.Fetch(p_Module_Name, 0, 0, 1, 0);
                Csla.Rules.BusinessRules.AddRule(
                              typeof(T),
                              new Csla.Rules.CommonRules.IsInRole(
                                Csla.Rules.AuthorizationActions.DeleteObject,
                               strRoles));

                //Load Edit rights
                objList = _dataPortal.Fetch(p_Module_Name, 0, 0, 0, 1);
                Csla.Rules.BusinessRules.AddRule(
                              typeof(T),
                              new Csla.Rules.CommonRules.IsInRole(
                                Csla.Rules.AuthorizationActions.EditObject,
                               strRoles));


               
            }

        }


        [Fetch]
        private void Fetch([Inject] ITableRightsDal dal, [Inject] IChildDataPortal<TableRightsInfo> childDataPortal)
        {
            Fetch(null, null, dal, childDataPortal);
        }
        [Fetch]
        private void Fetch(string tableName, string rolename, [Inject] ITableRightsDal dal, [Inject] IChildDataPortal<TableRightsInfo> childDataPortal)
        {
            using (LoadListMode)
            {
                List<TableRightsDto> list = null;
                if (string.IsNullOrEmpty(tableName) && string.IsNullOrEmpty(rolename))
                    list = dal.Fetch();
                else if (string.IsNullOrEmpty(tableName) && !string.IsNullOrEmpty(rolename))
                    list = dal.FetchforRoleName(rolename);
                else if (string.IsNullOrEmpty(rolename) && !string.IsNullOrEmpty(tableName))
                    list = dal.FetchforTableName(tableName);

                foreach (var item in list)
                    Add(childDataPortal.FetchChild(item));
            }
        }


        [Fetch]
        private void Fetch(string tableName, int canView, int canAdd,
            int canDelete, int canEdit,[Inject] ITableRightsDal dal)
        {
            using (LoadListMode)
            {
                //List<string> strRoles2 = new List<string>();
                //   List<Dal.TableRightsDto> list = null; 
                // list = dal.FetchforTableName(tableName);
                if (canView > 0)
                    strRoles = dal.FetchViewRoles(tableName);
                else if (canAdd > 0)
                    strRoles = dal.FetchAddRoles(tableName);
                else if (canDelete > 0)
                    strRoles = dal.FetchDeleteRoles(tableName);
                else if (canEdit > 0)
                    strRoles = dal.FetchEditRoles(tableName);

                //strRoles1 =  strRoles2 ;
                //  foreach (var item in list)
                //   Add(DataPortal.FetchChild<TableRightsInfo>(item));
            }
        }

    }
}
